### ISSUE TITLE
Transfer learning

### ISSUE THREAD

**Issue by tatiana-boura** [eneribnk/Tasks_plasmAI#6]
- [x] Establish best architectures for both datasets ($D_1$, $D_2$), via tuning.
- [x] Train 2 ML ( $M_1$,  $M_2$) models with their  optimal architectures and hyperparameters and report the results.
- [ ] The following experiments should be executed for both { $D_1$,  $M_2$ } and { $D_2$,  $M_1$ }
    - [x] As a baseline we should have how the pre-trained $M_1$ performs on $D_2$, **without any re-training on the new data**
    - [x] Then, take the trained $M_1$ and continue training it with the whole  $D_2$, using a **small** learning rate (*at least* 10 times smaller that the initial) and make sure the batch size is not small
    - [x] Obtain the trained network $M_{12}$
    - [x] We would  like for the  $M_{12}$ to perform better than $M_2$ on  $D_2$, as that whould indicate that the additional knowledge yielded better results
     - [x] However, we are at least aiming for the same results, but it is important that we make sure that the NN does not suffer from **catastrophic forgetting**. We check that by evaluating the $M_{12}$ on $D_1$ and expect to have results that are better that an untrained $M_{1_{untrained}}$ and worse than $M_{1}$.
    - [x] Perform transfer learning with **freezed layers** (this will help us make assumptions about the geometry of the problem that the network learns). In NNs, it is expected that earlier layers contain basic domain information and low-level features, so in transfer learning tasks we often freeze the initial layers and train the latter ones. Specifically:
      1. Just freeze the first layers and continue training the latter ones. Then do the following, 
      2. For an initial number of epochs freeze completely the first layers. The amount of layers to freeze and the initial number of epochs is decided upon experimentation.
      3. Then gradually, start un-freezing the layers backwards until the whole network is receives gradient training.
     - [x] I suspect that since we do a kind-off backwards transfer learning (as $M_{11}$ and  $M_{12}$ have different inputs yet similar outputs and not the other way around), the geometry may be inverted, so we should also perform the freezing backwards, i.e., freeze the last layers and re-train the first ones, or just train the middle ones.
     - [x] After performing these experiments, we shall iteratively use only a percentage of the $D_2$ and evaluate how the  $M_{12_{perc}}$ performs. I suggest trying 80%, 60%, 40%, 20%. We shall decide what version of the $M_{12}$ to use upon evaluating the aforementioned experiments.

Important things to note:

- The normalization of the values should be done using the statistics for the initially trained task, e.g. to train $M_{12}$ on $D_2$, $D_2$ should be normalized with the $\mu, \sigma$ of $D_1$. We will discuss this in detail. (revised: make sure we fit the dustributions at testing time)
- In all experiments keep the training and validation sets the same!
- If catastrophic forgetting occurs we will employ methods to avoid it, e.g. use the olda dataset after a few epochs alone or concatenated with the new data
- Things will get complicated so make sure to store all results in a easily navigatable manner
